_stbt/state_watch.py

```diff
@@ -0,0 +1,243 @@
+                        "file": "tests/my_file.py",
+                        "line": 654
+                    }
+                }
+            }
+        }
+
+    This record indicates that the current line changed to line 654 of file
+    "tests/my_file.py" at 20:55:26.0992343 UTC on 2014-11-28.  Notes:
+
+    * The root object has a single key "state_change".  This allows extensions
+      in the future for other types of messages to be sent over the same
+      protocol.
+
+    * The state_change message has two keys:
+
+        * "time" which is the time of the change using mongodb style date
+          format `{"$date": "2014-11-28T20:55:26.092343Z"}`.  The date timestamp
+          is in ISO8601 format.  The fact that this is an object rather than
+          just a string may seem a little odd but JSON has no native datetime
+          type so this is necessary to specify the type.
+
+        * "changes" is a dictionary of changes made to the state of the system.
+          The keys identify the value that is changing, and the values are the
+          new value of the respective key.  The key is dot seperated strings
+          identifying the subtree of the heirarical data structure that should
+          be replaced.  e.g:
+
+              "test_run.current_line": {"cow": "moo"}
+
+          means:
+
+              date["test_run"]["current_line"] = {"cow": "moo"}
+
+          The fact that multiple values may be replaced in one message allows
+          atomic changes to happen to the hierarchy.
+    """
+    def __init__(self, file_):
+        self._file = file_
+
+    def set(self, items, time=None):
+        """
+        >>> sw = StateSender(StringIO())
+        >>> sw.set({"animals.noises": {"cow": "moo"}})
+        """
+        if time is None:
+            time = datetime.datetime.now()
+        message = {
+            "state_change": {
+                "time": time.isoformat(),
+                "changes": items
+            }
+        }
+        self._file.write(json.dumps(message, sort_keys=True) + '\r\n')
+
+    def close(self):
+        self._file.close()
+        self._file = None
+
+    def log_test_starting(self, name, file_, function, line):
+        self.set({"test_run": {
+            "current_line": {},
+            "test_case": {
+                "name": name,
+                "file": file_,
+                "function": function,
+                "line": line
+            }}})
+
+    def log_test_ended(self):
+        self.set({"test_run": {}})
+
+    def log_current_line(self, file_, line):
+        self.set({"test_run.current_line": {"file": file_, "line": line}})
+
+
+def test_state_changes():
+    f = StringIO()
+    sw = StateSender(f)
+    sw.set({"test_run.line_number": 23},
+           time=datetime.datetime(2014, 3, 4, 12, 45, 12))
+    assert f.getvalue() == (
+        '{"state_change": {"changes": {"test_run.line_number": 23}, '
+        '"time": "2014-03-04T12:45:12"}}\r\n')
```

Comment:
This doesn't match the StateSender docstring which says that the timestamp is in mongo format (`"time": {"$date": ...}`).